### PR TITLE
My solution to the ruby exercise

### PR DESCRIPTION
**Pull request by fsperling**
The bot responds to mentions to @fxsperling.
If the tweet contains: !w term
it looks up the term on wikipedia and replies with the beginning of the requested wiki page.
It checks for mentions every hour triggered by the heroku scheduler or on request:
http://geckobot.herokuapp.com/controlbot

### CODE REVIEW

File: app/helpers/twitter_api_helper.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+module TwitterApiHelper 
+  require 'wikipedia'
+
+  def TwitterApiHelper.initClient
+    client = Twitter::REST::Client.new do |config|
+      config.consumer_key        = "vnZffQpHom39ewCmM6xkk1Ve6"
```
Comment (lcpriest) [inline app/helpers/twitter_api_helper.rb]: I recommend revoking these keys and then putting them into a .env file

File: app/helpers/twitter_api_helper.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      config.consumer_secret     = "Ftc7XwTndlbrbc3s71VfF2G4OSgdZRicsN971nZMV1AIHW0i37"
+      config.access_token        = "3791732720-wtRG6G14r7Au6o37HVZJgmrYOJdYdksriPmLBix"
+      config.access_token_secret = "yGluyaOq51LbBjzvtW1r2EAWEnyJb3bFzyk8pZsVf4XfH"
+    end
+    client
+  end
+
+  def TwitterApiHelper.answerToMentions
```
Comment (lcpriest) [inline app/helpers/twitter_api_helper.rb]: self.answer_to_mentions? Ruby style is underscored and self is all you need here.

File: app/helpers/twitter_api_helper.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    client = initClient
+    lastMentions = client.mentions_timeline(:count => 10)
+    activity = [] 
+
+    lastMentions.each do |mention|
+      unless Answeredmention.where(tweetid: mention.id).exists?
+        text = mention.text
+        sender = mention.user.screen_name    
+        activity.push({:sender => sender, :text => text})
```
Comment (lcpriest) [inline app/helpers/twitter_api_helper.rb]: Recommended hash syntax is now: 
```
{ sender: sender }
```

File: test/models/answeredmention_test.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class AnsweredmentionTest < ActiveSupport::TestCase
```
Comment (lcpriest) [inline test/models/answeredmention_test.rb]: Tests!

File: app/helpers/twitter_api_helper.rb (hunk 1 above)
Comment (lcpriest) [inline app/helpers/twitter_api_helper.rb]: As I now have access to this account

**Comment by lcpriest** [comment]
No major issues, your code style is not yet Ruby like, but that's an easy fix.
Tests are super important!

File: app/helpers/twitter_api_helper.rb (hunk 1 above)
Comment (fsperling) [inline app/helpers/twitter_api_helper.rb]: I know that should stay secret - I wasn't sure if you might want to test it by running it on your machine.
I moved the secrets out of the repo now.

File: app/helpers/twitter_api_helper.rb (hunk 2 above)
Comment (fsperling) [inline app/helpers/twitter_api_helper.rb]: I changed it.

File: app/helpers/twitter_api_helper.rb (hunk 3 above)
Comment (fsperling) [inline app/helpers/twitter_api_helper.rb]: Good tip.

**Comment by fsperling** [comment]
Hi Lachlan,

On 08.10.2015 11:45, Lachlan Priest wrote:
> No major issues, your code style is not yet Ruby like, but that's an 
> easy fix. Tests are super important!

Thank you for the feedback!

I agree tests are very important.
I'll add some if I find the time in the next days.

Felix

File: spec/lib/twitter_api_spec.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
+require 'spec_helper'
+require 'twitter_api'
+
+describe TwitterApi do
+  let!(:twitterapi) { TwitterApi.new } 
+
+  describe "init" do
+    it "should setup a twitter client" do
+      expect(twitterapi.init_client).to_not be_nil
```
Comment (lcpriest) [inline spec/lib/twitter_api_spec.rb]: I would also test that is an instance of TwitterApi, though this test is a little unnecessary as everything is done in the test suite anyway

File: app/models/answeredmention.rb [hunk 6]
```diff
@@ -0,0 +1,3 @@
+class Answeredmention < ActiveRecord::Base
```
Comment (lcpriest) [inline app/models/answeredmention.rb]: AnsweredMention

File: app/models/answeredmention.rb [hunk 7]
```diff
@@ -0,0 +1,3 @@
...
+  validates :tweetid, presence: true
```
Comment (lcpriest) [inline app/models/answeredmention.rb]: tweet_id

File: config/routes.rb [hunk 8]
```diff
@@ -5,6 +5,9 @@
   # You can have the root of your site routed with "root"
   root 'robots#index'
```
Comment (lcpriest) [inline config/routes.rb]: `resource :control_bot, only: :index`
`resource :replies, only: :create`
`resource :mentions, only: :index`

File: config/routes.rb [hunk 9]
```diff
@@ -5,6 +5,9 @@
...
 
+  match '/controlbot',          to: 'control_bot#view',  via: 'get'
+  match '/reply',               to: 'control_bot#reply',  via: 'get'
```
Comment (lcpriest) [inline config/routes.rb]: Can you explain the logic behind this naming?

File: lib/twitter_api.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      config.consumer_key        = ENV['TWITTER_CONSUMER_KEY'] 
+      config.consumer_secret     = ENV['TWITTER_CONSUMER_SECRET']
+      config.access_token        = ENV['TWITTER_ACCESS_TOKEN']
+      config.access_token_secret = ENV['TWITTER_ACCESS_TOKEN_SECRET']
+    end
+  end
+
+  def answer_to_mentions
+    client = init_client
+    lastMentions = client.mentions_timeline(:count => 10)
```
Comment (lcpriest) [inline lib/twitter_api.rb]: `lastMentions = client.mentions_timeline(count: 10)`

File: lib/twitter_api.rb [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+        client.update(reply)
+        Answeredmention.new(tweetid: mention.id).save
+      end
+    end 
+
+    activity
+  end
+
+  def process_mention(sender, text)
```
Comment (lcpriest) [inline lib/twitter_api.rb]: I recommend `rescue`ing this function also, all API requests should be dealt with in case of connection issues.

Comment (fsperling) [inline lib/twitter_api.rb]: I added rescue blocks for the API requests. And also renamed tweetid and Answeredmention.

File: config/routes.rb (hunk 9 above)
Comment (fsperling) [inline config/routes.rb]: Well, the robot performs certain actions triggered by cron. To test the bot I wanted to be able to manually trigger the action by calling an URL. So it's more function-like than RESTful. I think:
/bot/reply
/bot/searchstream
would have been better.

I think the checking for mentions and replying doesn't fit so well with REST as it is kind of a batch processing job.
But you could make the results of the bots work accessible via REST:

/bots/123/replies/1
/bots/123/replies/2
/bots/123/mentions/1
...

**Comment by lcpriest** [comment]
Thanks! Bradley should email you today.

**Comment by fsperling** [comment]
Hey,

On 06.11.2015 11:32, Lachlan Priest wrote:
> Thanks! Bradley should email you today.

Still haven't heard from him. Could you please remind him.

F

**Pull request closed without merging**